dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,519 @@
+            sleep(1.0)
+            # XXX: should we add some break condition and raise exception?
+            return False
+
+    def upload_file(self, from_info, to_info, progress_name):
+        """Implements resumable upload protocol
+
+        https://developers.google.com/drive/api/v3/manage-uploads#resumable
+        """
+
+        dirname = posixpath.dirname(to_info.path).strip("/")
+        if dirname:
+            parent = self.makedirs(to_info.root, dirname)
+        else:
+            parent = to_info.root
+
+        # initiate resumable upload
+        resumable_upload_url = self._resumable_upload_initiate(
+            parent, posixpath.basename(to_info.path)
+        )
+
+        from_file = open(from_info.path, "rb")
+        if progress_name is not None:
+            from_file = track_progress(progress_name, from_file)
+
+        file_size = os.fstat(from_file.fileno()).st_size
+
+        success = self._resumable_upload_first_request(
+            resumable_upload_url, from_file, to_info, file_size
+        )
+        while not success:
+            success = self._resumable_upload_resume(
+                resumable_upload_url, from_file, to_info, file_size
+            )
+
+    def upload(self, from_infos, to_infos, names=None, no_progress_bar=False):
+
+        names = self._verify_path_args(to_infos, from_infos, names)
+
+        for from_info, to_info, name in zip(from_infos, to_infos, names):
+
+            if from_info.scheme != Schemes.LOCAL:
+                raise NotImplementedError
+
+            if to_info.scheme != self.scheme:
+                raise NotImplementedError
+
+            logger.debug(
+                "Uploading '{}' to '{}/{}'".format(
+                    from_info.path, self.url, to_info.path
+                )
+            )
+
+            if not name:
+                name = os.path.basename(from_info.path)
+
+            if not no_progress_bar:
+                progress.update_target(name, 0, None)
+
+            try:
+                self.upload_file(
+                    from_info,
+                    to_info,
+                    progress_name=name if no_progress_bar is False else None,
+                )
+            except Exception:
+                msg = "failed to upload '{}' to '{}/{}'"
+                logger.exception(
+                    msg.format(from_info.path, self.url, to_info.path)
+                )
+                continue
+
+            progress.finish_target(name)
+
+    def download_file(self, from_info, to_info, progress_name=None):
+        metadata = self.get_metadata_by_path(
+            from_info.root, from_info.path, fields=["id", "mimeType", "size"]
+        )
+        response = self.request(
+            "GET",
+            "drive/v3/files/" + metadata["id"],
+            params={"alt": "media"},
+            stream=True,
+        )
+        current = 0
+        if response.status_code != 200:
+            try:
```

Comment:
the whole block worth making an utils function. And please let's not catch Exception.

Reply:
Thanks!